### PR TITLE
feat(storage): pass epoch and table id before barrier

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Pass the table_ids that will start writing on an epoch at the beginning of `inject_barrier`. This is for helping HummockUploader to spill data correctly so that data of two separate syncs won't be mixed in a spill task.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/barrier/mod.rs [hunk 1]
```diff
@@ -766,10 +766,10 @@ impl GlobalBarrierManager {
 
         send_latency_timer.observe_duration();
 
-        let node_to_collect = match self
-            .control_stream_manager
-            .inject_barrier(command_ctx.clone())
-        {
+        let node_to_collect = match self.control_stream_manager.inject_barrier(
+            command_ctx.clone(),
+            self.state.inflight_actor_infos.existing_table_ids(),
```
Comment (hzxa21) [inline src/meta/src/barrier/mod.rs:771]: Just to confirm, `self.state.inflight_actor_infos.existing_table_ids()` contains both created MV's and creating MV's state table ids, right?

File: src/storage/src/hummock/event_handler/uploader/mod.rs [hunk 2]
```diff
@@ -727,7 +758,7 @@ struct UnsyncData {
     instance_table_id: HashMap<LocalInstanceId, TableId>,
     // TODO: this is only used in spill to get existing epochs and can be removed
     // when we support spill not based on epoch
-    epochs: BTreeMap<HummockEpoch, ()>,
+    epochs: BTreeMap<HummockEpoch, HashSet<TableId>>,
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader/mod.rs:761]: Seems to be conflicting with https://github.com/risingwavelabs/risingwave/pull/17539

File: src/storage/src/hummock/store/hummock_storage.rs [hunk 3]
```diff
@@ -454,6 +454,15 @@ impl HummockStorage {
         )
     }
 
+    /// Declare the start of an epoch. This information is provided for spill so that the spill task won't
+    /// include data of two or more syncs.
+    // TODO: remote this method when we support spill task that can include data of more two or more syncs
```
Comment (hzxa21) [inline src/storage/src/hummock/store/hummock_storage.rs:459]: typo: `remove`

File: src/storage/src/hummock/event_handler/uploader/mod.rs [hunk 4]
```diff
@@ -807,7 +826,20 @@ impl UnsyncData {
             .get_mut(&instance_id)
             .expect("should exist");
         let epoch = instance_data.local_seal_epoch(next_epoch);
-        self.epochs.insert(next_epoch, ());
+        // When drop/cancel a streaming job, for the barrier to stop actor, the
+        // local instance will call `local_seal_epoch`, but the `next_epoch` won't be
+        // called `start_epoch` because we have stopped writing on it.
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader/mod.rs:831]: Does it mean that the actors for the dropping/cancelling streaming job will be included in `actors_to_send` but the table_ids for the streaming job will not be included in the barrier?

File: src/storage/src/hummock/event_handler/uploader/mod.rs [hunk 5]
```diff
@@ -807,7 +826,20 @@ impl UnsyncData {
...
+        if !table_data.unsync_epochs.contains_key(&next_epoch) {
+            if let Some(stopped_next_epoch) = table_data.stopped_next_epoch {
+                assert_eq!(stopped_next_epoch, next_epoch);
+            } else {
+                if let Some(max_epoch) = table_data.max_epoch() {
+                    assert_gt!(next_epoch, max_epoch);
+                }
+                debug!(?table_id, epoch, next_epoch, "table data has stopped");
+                table_data.stopped_next_epoch = Some(next_epoch);
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader/mod.rs:840]: so `stopped_next_epoch` is introduced only for assertion check?

File: src/storage/src/hummock/event_handler/uploader/mod.rs [hunk 6]
```diff
@@ -807,7 +826,20 @@ impl UnsyncData {
             .get_mut(&instance_id)
             .expect("should exist");
         let epoch = instance_data.local_seal_epoch(next_epoch);
-        self.epochs.insert(next_epoch, ());
+        // When drop/cancel a streaming job, for the barrier to stop actor, the
+        // local instance will call `local_seal_epoch`, but the `next_epoch` won't be
+        // called `start_epoch` because we have stopped writing on it.
+        if !table_data.unsync_epochs.contains_key(&next_epoch) {
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader/mod.rs:832]: is it possible that `table_data` for the dropping/cancelling streaming job tables are absent causing L823 to panic?

File: src/meta/src/barrier/mod.rs (hunk 1 above)
Comment (wenym1) [reply src/meta/src/barrier/mod.rs:771]: Yes. This is more like a refactoring PR, and the logic of partial checkpoint based backfill is not included yet.

File: src/storage/src/hummock/event_handler/uploader/mod.rs (hunk 4 above)
Comment (wenym1) [reply src/storage/src/hummock/event_handler/uploader/mod.rs:831]: Yes. This is the key difference between `table_ids_to_sync` and `actors_to_collect` in the `InjectBarrierRequest`. `actors_to_collect` means actors that this barrier will flow through, and the dropping streaming job will stop after the barrier flow through the actor. 

On the other hand, `table_ids_to_sync` means the table id that will start writing data on the `epoch.curr` after the barrier, so it won't include the table_ids of dropping streaming job. For this reason, in `LocalBarrierWorker`, we store the `table_ids_to_sync` of the previous barrier (let's say epoch{curr=2, prev=1}), and when a barrier(epoch{curr=3, prev=2}) flows through the streaming graph, it means the table id in the `table_ids_to_sync` of the previous barrier(epoch{curr=2, prev=1}) has finished writing the data of epoch 2, and then we call `sync` with the `table_ids_to_sync` of the previous barrier.

File: src/storage/src/hummock/event_handler/uploader/mod.rs (hunk 5 above)
Comment (wenym1) [reply src/storage/src/hummock/event_handler/uploader/mod.rs:840]: Yes.

File: src/storage/src/hummock/event_handler/uploader/mod.rs (hunk 6 above)
Comment (wenym1) [reply src/storage/src/hummock/event_handler/uploader/mod.rs:832]: It's unlikely to happen. 

Let's say the barrier that stops the actor has epoch{curr=3, prev=2}. Before calling `sync` on `epoch.prev`, we must have called `start_epoch(epoch.prev)`. Before calling this `local_seal_epoch(epoch.curr)`, this barrier won't be collected, and we cannot call `sync(epoch.prev)`. Before `sync(epoch.prev)`, the `unsync_epochs` must contain `epoch.prev`, and therefore the `table_data` is non-empty and won't be dropped, and then in L823 the table won't be absent.

Comment (hzxa21) [reply src/storage/src/hummock/event_handler/uploader/mod.rs:832]: Got it. Thanks for the explanation.

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**